dj/sql/parsing/backends/sqloxide.py

```diff
@@ -0,0 +1,391 @@
+"""
+parsing backend turing sqloxide output into DJ AST
+"""
+from typing import List, Set, Union, cast
+
+from sqloxide import parse_sql
+
+from dj.sql.parsing.ast import (
+    Alias,
+    Between,
+    BinaryOp,
+    BinaryOpKind,
+    Boolean,
+    Case,
+    Column,
+    Expression,
+    From,
+    Function,
+    Join,
+    JoinKind,
+    Number,
+    Operation,
+    Query,
+    Select,
+    String,
+    Table,
+    UnaryOp,
+    UnaryOpKind,
+    Value,
+    Wildcard,
+)
+from dj.sql.parsing.backends.exceptions import DJParseException
+
+
+def match_keys(parse_tree: dict, *keys: Set[str]) -> bool:
+    """
+    match a dictionary having exact keys
+    """
+    return set(parse_tree.keys()) in keys
+
+
+def match_keys_subset(parse_tree: dict, *keys: Set[str]) -> bool:
+    """
+    match a dict having a subset of keys
+    """
+    tree_keys = set(parse_tree.keys())
+    return any(key <= tree_keys for key in keys)  # pragma: no cover
```

Comment:
Why no unit test coverage here? I can understand the Exception cases below, but not sure about this one.

Reply:
great question. I think the coverage is confused here. That sounds a little crazy but I'll try to explain what I think is happening and hopefully you can correct me. (and @samredai  explained some of what coverage might be thinking here as well so perhaps he can add).

The coverage missing statement was "47->exit" suggesting the function may exit early at that line. It's not clear there is any branching. The goal with this line was to allow for early stopping since the `any` can lazily go through the generator comprehension and just stop at the first `True`. As @samredai suggested to me, the coverage may be taking into account the "early stopping branching" logic. I ignored coverage here as if that's the case it seemed irrelevant.

Reply:
That makes sense. I haven't seen this kind of case in a while :)

Reply:
Yeah I'm pretty sure that's exactly what's happening here since if you cast the generator to a tuple, causing the generator to iterate to completion, the branching coverage miss goes away. I learned something new here, coverage is much smarter than I thought!